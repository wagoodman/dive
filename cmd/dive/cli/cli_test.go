package cli

import (
	"bytes"
	"github.com/anchore/clio"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/google/shlex"
	"github.com/spf13/cobra"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"io"
	"os"
	"testing"
)

// TODO: remove coverage from consideration of these tests (or make different test classes)

func Test_Config(t *testing.T) {
	t.Setenv("DIVE_CONFIG", "./testdata/image-multi-layer-dockerfile/dive.yaml")

	rootCmd := getTestCommand(t, "config --load")
	all := Capture().All().Run(t, func() {
		require.NoError(t, rootCmd.Execute())
	})

	snaps.MatchSnapshot(t, all)
}

func Test_LegacyRules(t *testing.T) {
	t.Setenv("DIVE_CONFIG", "./testdata/config/dive-ci-legacy.yaml")

	rootCmd := getTestCommand(t, "config --load")
	all := Capture().All().Run(t, func() {
		require.NoError(t, rootCmd.Execute())
	})

	assert.Contains(t, all, "lowest-efficiency: '0.95'", "missing lowest-efficiency legacy rule")
	assert.Contains(t, all, "highest-wasted-bytes: '20MB'", "missing highest-wasted-bytes legacy rule")
	assert.Contains(t, all, "highest-user-wasted-percent: '0.2'", "missing highest-user-wasted-percent legacy rule")
}

func Test_Build_Dockerfile(t *testing.T) {
	t.Setenv("DIVE_CONFIG", "./testdata/image-multi-layer-dockerfile/dive.yaml")

	t.Run("implicit dockerfile", func(t *testing.T) {
		rootCmd := getTestCommand(t, "build testdata/image-multi-layer-dockerfile")
		stdout := Capture().WithStdout().WithSuppress().Run(t, func() {
			require.NoError(t, rootCmd.Execute())
		})
		snaps.MatchSnapshot(t, stdout)
	})

	t.Run("explicit file flag", func(t *testing.T) {
		rootCmd := getTestCommand(t, "build testdata/image-multi-layer-dockerfile -f testdata/image-multi-layer-dockerfile/Dockerfile")
		stdout := Capture().WithStdout().WithSuppress().Run(t, func() {
			require.NoError(t, rootCmd.Execute())
		})
		snaps.MatchSnapshot(t, stdout)
	})
}

func Test_Build_Containerfile(t *testing.T) {
	t.Setenv("DIVE_CONFIG", "./testdata/image-multi-layer-containerfile/dive.yaml")

	t.Run("implicit containerfile", func(t *testing.T) {
		rootCmd := getTestCommand(t, "build testdata/image-multi-layer-containerfile")
		stdout := Capture().WithStdout().WithSuppress().Run(t, func() {
			require.NoError(t, rootCmd.Execute())
		})
		snaps.MatchSnapshot(t, stdout)
	})

	t.Run("explicit file flag", func(t *testing.T) {
		rootCmd := getTestCommand(t, "build testdata/image-multi-layer-containerfile -f testdata/image-multi-layer-containerfile/Containerfile")
		stdout := Capture().WithStdout().WithSuppress().Run(t, func() {
			require.NoError(t, rootCmd.Execute())
		})
		snaps.MatchSnapshot(t, stdout)
	})
}

func getTestCommand(t testing.TB, cmd string) *cobra.Command {
	c := Command(clio.Identification{
		Name:    "dive",
		Version: "testing",
	})

	args, err := shlex.Split(cmd)
	require.NoError(t, err, "failed to parse command line %q", cmd)

	c.SetArgs(args)

	return c
}

type capturer struct {
	stdout   bool
	stderr   bool
	suppress bool
}

func Capture() *capturer {
	return &capturer{}
}

func (c *capturer) WithSuppress() *capturer {
	c.suppress = true
	return c
}

func (c *capturer) All() *capturer {
	c.stdout = true
	c.stderr = true
	return c
}

func (c *capturer) WithStdout() *capturer {
	c.stdout = true
	return c
}

func (c *capturer) WithStderr() *capturer {
	c.stderr = true
	return c
}

func (c *capturer) Run(t testing.TB, f func()) string {
	t.Helper()

	r, w, err := os.Pipe()
	if err != nil {
		panic(err)
	}

	devNull, err := os.OpenFile(os.DevNull, os.O_WRONLY, 0)
	if err != nil {
		panic(err)
	}
	defer devNull.Close()

	oldStdout := os.Stdout
	oldStderr := os.Stderr

	if c.stdout {
		os.Stdout = w
	} else if c.suppress {
		os.Stdout = devNull
	}

	if c.stderr {
		os.Stderr = w
	} else if c.suppress {
		os.Stderr = devNull
	}

	defer func() {
		os.Stdout = oldStdout
		os.Stderr = oldStderr
	}()

	f()
	require.NoError(t, w.Close())

	var buf bytes.Buffer
	_, err = io.Copy(&buf, r)
	require.NoError(t, err)

	return buf.String()
}
