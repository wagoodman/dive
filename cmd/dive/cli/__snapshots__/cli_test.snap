
[Test_Config - 1]
log:
  # suppress all logging output (env: DIVE_LOG_QUIET)
  quiet: false

  # explicitly set the logging level (available: [error warn info debug trace]) (env: DIVE_LOG_LEVEL)
  level: 'warn'

  # file path to write logs to (env: DIVE_LOG_FILE)
  file: ''

dev:
  # capture resource profiling data (available: [cpu, mem]) (env: DIVE_DEV_PROFILE)
  profile: 'none'

# container engine to use for image analysis (supported options: 'docker' and 'podman') (env: DIVE_CONTAINER_ENGINE)
container-engine: 'docker'

# continue with analysis even if there are errors parsing the image archive (env: DIVE_IGNORE_ERRORS)
ignore-errors: false

# enable CI mode (env: DIVE_CI)
ci: true

# path to the CI config file (env: DIVE_CI_CONFIG)
ci-config: '.dive-ci'

rules:
  # lowest allowable image efficiency (as a ratio between 0-1), otherwise CI validation will fail. (env: DIVE_RULES_LOWEST_EFFICIENCY)
  lowest-efficiency: '0.9'

  # highest allowable bytes wasted, otherwise CI validation will fail. (env: DIVE_RULES_HIGHEST_WASTED_BYTES)
  highest-wasted-bytes: '20MB'

  # highest allowable percentage of bytes wasted (as a ratio between 0-1), otherwise CI validation will fail. (env: DIVE_RULES_HIGHEST_USER_WASTED_PERCENT)
  highest-user-wasted-percent: '0.90'

# Skip the interactive TUI and write the layer analysis statistics to a given file. (env: DIVE_JSON_PATH)
json-path: ''

keybinding:
  # quit the application (global) (env: DIVE_KEYBINDING_QUIT)
  quit: 'ctrl+c'

  # toggle between different views (global) (env: DIVE_KEYBINDING_TOGGLE_VIEW)
  toggle-view: 'tab'

  # filter files by name (global) (env: DIVE_KEYBINDING_FILTER_FILES)
  filter-files: 'ctrl+f, ctrl+slash'

  # close file filtering (global) (env: DIVE_KEYBINDING_CLOSE_FILTER_FILES)
  close-filter-files: 'esc'

  # move cursor up (global) (env: DIVE_KEYBINDING_UP)
  up: 'up,k'

  # move cursor down (global) (env: DIVE_KEYBINDING_DOWN)
  down: 'down,j'

  # move cursor left (global) (env: DIVE_KEYBINDING_LEFT)
  left: 'left,h'

  # move cursor right (global) (env: DIVE_KEYBINDING_RIGHT)
  right: 'right,l'

  # scroll page up (file view) (env: DIVE_KEYBINDING_PAGE_UP)
  page-up: 'pgup,u'

  # scroll page down (file view) (env: DIVE_KEYBINDING_PAGE_DOWN)
  page-down: 'pgdn,d'

  # compare all layers (layer view) (env: DIVE_KEYBINDING_COMPARE_ALL)
  compare-all: 'ctrl+a'

  # compare specific layer (layer view) (env: DIVE_KEYBINDING_COMPARE_LAYER)
  compare-layer: 'ctrl+l'

  # toggle directory collapse (file view) (env: DIVE_KEYBINDING_TOGGLE_COLLAPSE_DIR)
  toggle-collapse-dir: 'space'

  # toggle collapse all directories (file view) (env: DIVE_KEYBINDING_TOGGLE_COLLAPSE_ALL_DIR)
  toggle-collapse-all-dir: 'ctrl+space'

  # toggle visibility of added files (file view) (env: DIVE_KEYBINDING_TOGGLE_ADDED_FILES)
  toggle-added-files: 'ctrl+a'

  # toggle visibility of removed files (file view) (env: DIVE_KEYBINDING_TOGGLE_REMOVED_FILES)
  toggle-removed-files: 'ctrl+r'

  # toggle visibility of modified files (file view) (env: DIVE_KEYBINDING_TOGGLE_MODIFIED_FILES)
  toggle-modified-files: 'ctrl+m'

  # toggle visibility of unmodified files (file view) (env: DIVE_KEYBINDING_TOGGLE_UNMODIFIED_FILES)
  toggle-unmodified-files: 'ctrl+u'

  # toggle display of file attributes (file view) (env: DIVE_KEYBINDING_TOGGLE_FILETREE_ATTRIBUTES)
  toggle-filetree-attributes: 'ctrl+b'

  # toggle sort order (file view) (env: DIVE_KEYBINDING_TOGGLE_SORT_ORDER)
  toggle-sort-order: 'ctrl+o'

  # (env: DIVE_KEYBINDING_TOGGLE_WRAP_TREE)
  toggle-wrap-tree: 'ctrl+p'

  # extract file contents (file view) (env: DIVE_KEYBINDING_EXTRACT_FILE)
  extract-file: 'ctrl+e'

diff:
  # types of file differences to hide (added, removed, modified, unmodified) (env: DIVE_DIFF_HIDE)
  hide:
    - 'added'

filetree:
  # collapse directories by default in the filetree (env: DIVE_FILETREE_COLLAPSE_DIR)
  collapse-dir: true

  # percentage of screen width for the filetree pane (must be >0 and <1) (env: DIVE_FILETREE_PANE_WIDTH)
  pane-width: 0.7

  # show file attributes in the filetree view (env: DIVE_FILETREE_SHOW_ATTRIBUTES)
  show-attributes: false

layer:
  # show aggregated changes across all previous layers (env: DIVE_LAYER_SHOW_AGGREGATED_CHANGES)
  show-aggregated-changes: true

---

[Test_Build_Dockerfile/implicit_dockerfile - 1]
[1mAnalysis:[0m
  efficiency:        100.00 %
  wastedBytes:       131 bytes (131 B)
  userWastedPercent: 71.98 %
[1m
Inefficient Files:[0mCount  Wasted Space  File Path
    3          80 B  /root/saved.txt
    2          34 B  /root/example/somefile1.txt
    2          17 B  /root/example/somefile3.txt
    2           0 B  /root
   10           0 B  /etc
[1m
Evaluation:[0m
  PASS: highestUserWastedPercent
  PASS: highestWastedBytes
  PASS: lowestEfficiency
[32mResult:PASS [Total:3] [Passed:3] [Failed:0] [Warn:0] [Skipped:0][0m


---

[Test_Build_Dockerfile/explicit_file_flag - 1]
[1mAnalysis:[0m
  efficiency:        100.00 %
  wastedBytes:       131 bytes (131 B)
  userWastedPercent: 71.98 %
[1m
Inefficient Files:[0mCount  Wasted Space  File Path
    3          80 B  /root/saved.txt
    2          34 B  /root/example/somefile1.txt
    2          17 B  /root/example/somefile3.txt
    2           0 B  /root
   10           0 B  /etc
[1m
Evaluation:[0m
  PASS: highestUserWastedPercent
  PASS: highestWastedBytes
  PASS: lowestEfficiency
[32mResult:PASS [Total:3] [Passed:3] [Failed:0] [Warn:0] [Skipped:0][0m


---

[Test_Build_Containerfile/implicit_containerfile - 1]
[1mAnalysis:[0m
  efficiency:        100.00 %
  wastedBytes:       131 bytes (131 B)
  userWastedPercent: 71.98 %
[1m
Inefficient Files:[0mCount  Wasted Space  File Path
    3          80 B  /root/saved.txt
    2          34 B  /root/example/somefile1.txt
    2          17 B  /root/example/somefile3.txt
    2           0 B  /root
   10           0 B  /etc
[1m
Evaluation:[0m
  PASS: highestUserWastedPercent
  PASS: highestWastedBytes
  PASS: lowestEfficiency
[32mResult:PASS [Total:3] [Passed:3] [Failed:0] [Warn:0] [Skipped:0][0m


---

[Test_Build_Containerfile/explicit_file_flag - 1]
[1mAnalysis:[0m
  efficiency:        100.00 %
  wastedBytes:       131 bytes (131 B)
  userWastedPercent: 71.98 %
[1m
Inefficient Files:[0mCount  Wasted Space  File Path
    3          80 B  /root/saved.txt
    2          34 B  /root/example/somefile1.txt
    2          17 B  /root/example/somefile3.txt
    2           0 B  /root
   10           0 B  /etc
[1m
Evaluation:[0m
  PASS: highestUserWastedPercent
  PASS: highestWastedBytes
  PASS: lowestEfficiency
[32mResult:PASS [Total:3] [Passed:3] [Failed:0] [Warn:0] [Skipped:0][0m


---
